pkg/container/docker_run.go

```diff
@@ -486,87 +487,126 @@ func (cr *containerReference) extractPath(env *map[string]string) common.Executo
 	}
 }
 
-func (cr *containerReference) exec(cmd []string, env map[string]string, user, workdir string) common.Executor {
-	return func(ctx context.Context) error {
-		logger := common.Logger(ctx)
-		// Fix slashes when running on Windows
-		if runtime.GOOS == "windows" {
-			var newCmd []string
-			for _, v := range cmd {
-				newCmd = append(newCmd, strings.ReplaceAll(v, `\`, `/`))
-			}
-			cmd = newCmd
+func (cr *containerReference) getWorkdir(workdir string) string {
+	if workdir != "" {
+		if strings.HasPrefix(workdir, "/") {
+			return workdir
 		}
+		return fmt.Sprintf("%s/%s", cr.input.WorkingDir, workdir)
+	}
+	return cr.input.WorkingDir
+}
 
-		logger.Debugf("Exec command '%s'", cmd)
-		isTerminal := term.IsTerminal(int(os.Stdout.Fd()))
-		envList := make([]string, 0)
-		for k, v := range env {
-			envList = append(envList, fmt.Sprintf("%s=%s", k, v))
+func getEnvListFromMap(env map[string]string) []string {
+	envList := make([]string, 0)
+	for k, v := range env {
+		envList = append(envList, fmt.Sprintf("%s=%s", k, v))
+	}
+	return envList
+}
+
+func (cr *containerReference) exec2(ctx context.Context, cmd []string, env map[string]string, user, workdir string) error {
+	logger := common.Logger(ctx)
+	// Fix slashes when running on Windows
+	if runtime.GOOS == "windows" {
+		var newCmd []string
+		for _, v := range cmd {
+			newCmd = append(newCmd, strings.ReplaceAll(v, `\`, `/`))
 		}
+		cmd = newCmd
+	}
 
-		var wd string
-		if workdir != "" {
-			if strings.HasPrefix(workdir, "/") {
-				wd = workdir
-			} else {
-				wd = fmt.Sprintf("%s/%s", cr.input.WorkingDir, workdir)
+	logger.Debugf("Exec command '%s'", cmd)
+	envList := getEnvListFromMap(env)
+	wd := cr.getWorkdir(workdir)
+	logger.Debugf("Working directory '%s'", wd)
+	idResp, err := cr.cli.ContainerExecCreate(ctx, cr.id, types.ExecConfig{
+		User:         user,
+		Cmd:          cmd,
+		WorkingDir:   wd,
+		Env:          envList,
+		Tty:          cr.input.Tty,
+		AttachStderr: true,
+		AttachStdout: true,
+		AttachStdin:  cr.input.Tty,
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	resp, err := cr.cli.ContainerExecAttach(ctx, idResp.ID, types.ExecStartCheck{
+		Tty: cr.input.Tty,
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer resp.Close()
+
+	var outWriter io.Writer
+	outWriter = cr.input.Stdout
+	if outWriter == nil {
+		outWriter = os.Stdout
+	}
+	errWriter := cr.input.Stderr
+	if errWriter == nil {
+		errWriter = os.Stderr
+	}
+
+	if cr.input.Tty {
+		// Avoid a freeze of programs expecting input like git asking for password and node
+		go func() {
+			c := 1
+			var err error
+			for c == 1 && err == nil {
+				c, err = resp.Conn.Write([]byte{4})
```

Comment:
Just to understand this code better: Is `c` the number of characters written? So this loop will write `EOT` every second, until it can't write anymore or am I missing something here?
Also, what exactly is "can't write anymore"? When the terminal is closed / the process has exited or is there something else happening?

Reply:
Yes it writes every second EOT until it fails to write, otherwise such a bash script will still stall the runner
```bash
#  Consumes first EOT
node
#  node stoped, the next node Consumes the second EOT
node
#  node stoped, the next node Consumes the third EOT
node
```
and so on...

Yes  "can't write anymore" Should be
> When the terminal is closed

In this case the process should be completed.

_I came to this implementation by trial and error, might be suboptimal.